Add test cases for granting and revoking data permissions

psarna opened the pull request:
This series adds the infrastructure needed for testing user permissions, like the ability to create temporary roles and CQL sessions which log in as different users, and a few initial test cases for granting and revoking permissions.

psarna commented:
Porting `--runveryslow` from alternator turned out to be trivial, so I did that. I'll finish the rest of the cases later, and likely all of them will be marked with `@pytest.mark.veryslow`.

nyh commented:
Before I review anything, just a comment on the "veryslow" mark and "--runveryslow": It is good for tests that have to be very slow, but the downside is that it means that these tests are **not** run by Jenkins, and in practice, may not be run by anyone after the initial development - although they will be easy to run if a bug is suspected, which is the primary goal of these tests.

Two techniques even better than the veryslow mark - if they fit your use case:

1. If the slowness happens because of a high timeout which is arbitrary, it can be made configurable. For example, to be able to test the Alternator TTL feature, test/alternator/run uses "`--alternator-ttl-period-in-seconds', '1'` so the TTL tests will finish in around 1 second, instead of 24 hours.

2. Sometimes if the tests is not super slow (e.g., just a few seconds), it might be possible to merge several tests into one, where a lot of tests happen in a single few-second duration (instead of many separate durations). An example is `test_tracing_all()` in `test/alternator/test_tracing.py`, where we test tracing of many Alternator operations in a single two-second sleep (which is necessary because of hard-code sleeps in Scylla's tracing implementation).

nyh commented on test/cql-pytest/util.py:
```diff
@@ -14,6 +14,10 @@
 
+from cassandra.auth import PlainTextAuthProvider
+from cassandra.cluster import Cluster, ConsistencyLevel, ExecutionProfile, EXEC_PROFILE_DEFAULT
+from cassandra.policies import RoundRobinPolicy
```
Most of these imports were prevbiously in conftest.py, so you can remove them from contest.py in this patch.

nyh commented on test/cql-pytest/conftest.py:
```diff
@@ -36,34 +36,12 @@ def pytest_addoption(parser):
-        # different versions. If we drop this setting completely, it will
-        # mean pick the latest version supported by the client and the server.
-        protocol_version=4,
-        # Use the default superuser credentials, which work for both Scylla and Cassandra
```
Maybe this comment is worth preserving in the copied code. It's not obvious that "cassandra" is a useruser account (and also on Scylla), hard-coded in the code. To be honest, I don't even remember why this is the case and why it's not considered a potential security hole.

psarna replied:
sure, restored

nyh commented on test/cql-pytest/util.py:
```diff
@@ -139,6 +143,50 @@ def new_secondary_index(cql, table, column, name='', extra=''):
+
+def new_session(cql, username):
+    endpoint = cql.hosts[0].endpoint
+    return cql_session(host=endpoint.address, port=endpoint.port, ssl=False, username=username, password=username)
```
Please make this or cql_session a contextmanager as well, and in the finally, do cluster.shutdown() when the session ends. Otherwise it leaves around threads and who knows what.
In the cql fixture I forgot to do it because there's just one of them, but in the check_cql() function in test/cql-pytest/run.py, I didn't forget. 
We should also fix the cql fixture to yield, not return, and then call cluster.shutdown().

nyh commented on test/cql-pytest/test_permissions.py:
```diff
@@ -39,3 +40,46 @@ def test_grant_applicable_data_and_role_permissions(cql, test_keyspace, cassandr
                     cql.execute(f"GRANT {permission} ON {resource} TO {user}")
+
+
+def eventually_authorized(fun, retries=100):
```
Nitpick (feel free to reject): I personally prefer to see a timeout in seconds, because otherwise there is always down when you see "retries=100" what it means, and you need to go calculating (in this case it's easy, but when the delay is exponential, it's not).

nyh commented on test/cql-pytest/test_permissions.py:
```diff
@@ -39,3 +40,46 @@ def test_grant_applicable_data_and_role_permissions(cql, test_keyspace, cassandr
+            return
+    try:
+        fun()
+        raise Exception(f"Function {fun} was not refused as unauthorized")
```
there's also pytest.fail() that perhaps more clearly indicates that this is a test failure, not a normal exception?

psarna replied:
missed this comment, I'll update in a few minutes

nyh commented on test/cql-pytest/test_permissions.py:
```diff
@@ -39,3 +40,46 @@ def test_grant_applicable_data_and_role_permissions(cql, test_keyspace, cassandr
+
+# Test that data permissions can be granted and revoked, and that they're effective
+# Unfortunately it takes ~3 seconds on Scylla and ~4 seconds on Cassandra to refresh
+# permissions, even though they're supposed to be refreshed on new login.
```
Did we know about this refresh time? Do we consider this acceptable - I see you used the phrasing "even though", maybe it is a bug?

Anyway, as I noted above I think it's fine that we have a test that takes 3 seconds. It will be less fine to have 10 or 20 of them. The question is whether we could merge several tests into one. For example, grant to username the ability to do 7 unrelated things, and then with one sleep, check all those 7 unrelated things. I'm not sure we can actually merge all the different things into one test - and that it will be as powerful as the separate tests. But maybe it will be useful to have a least one test without the veryslow marker that can be run by Jenkins.

psarna replied:
This is a priceless comment!!!!!!!! I didn't even suspect that this period might be configurable and I was just looking for the hardcoded value in the code, but it looks like it's configurable after all:
https://github.com/scylladb/scylla/blob/7e2d2cf1c18ddf471df1c2a28e805a5c1c69af50/db/config.cc#L726-L727

, both in Scylla and Cassandra!

That's fantastic news, because it makes the whole `runveryslow` thing redundant. I verified locally and indeed the tests run in absolutely acceptable times when this variable is set to `100ms`.

nyh reviewed (commented):
Looks good!
I left some comments about whether veryslow is the best thing we can do. The "veryslow" idea is great for development and debugging - it's like a manual test but it stays there in the code ready for the next time that you or someone else needs to check it again in the future. But the downside is that Jenkins doesn't run it, so it can't catch regressions in a timely matter. I raised a few ideas on how we can do the tests - or at least some of them - without veryslow.

psarna commented:
> Before I review anything, just a comment on the "veryslow" mark and "--runveryslow": It is good for tests that have to be very slow, but the downside is that it means that these tests are **not** run by Jenkins, and in practice, may not be run by anyone after the initial development - although they will be easy to run if a bug is suspected, which is the primary goal of these tests.
> 
> Two techniques even better than the veryslow mark - if they fit your use case:
> 
> 1. If the slowness happens because of a high timeout which is arbitrary, it can be made configurable. For example, to be able to test the Alternator TTL feature, test/alternator/run uses "`--alternator-ttl-period-in-seconds', '1'` so the TTL tests will finish in around 1 second, instead of 24 hours.

This is a priceless comment!!!!!!!! I didn't even suspect that this period might be configurable and I was just looking for the hardcoded value in the code, but it looks like it's configurable after all:
https://github.com/scylladb/scylla/blob/7e2d2cf1c18ddf471df1c2a28e805a5c1c69af50/db/config.cc#L726-L727

, both in Scylla and Cassandra!

That's fantastic news, because it makes the whole `runveryslow` thing redundant. I verified locally and indeed the tests run in absolutely acceptable times when this variable is set to `100ms`.

psarna commented:
v2:
 - instead of marking the tests as veryslow, servers are set up to refresh permissions significantly faster
 - the test case is upgraded to test all layers of data permissions: for all keyspaces, for a specific keyspace and for a specific table

psarna commented:
v3:
 - removed redundant imports from conftest.py
 - transformed throwing an exception to pytest.fail in eventually_unauthorized
 - restored a previously dropped comment on default username and password
 - changed the unit for retries from an arbitrary number to seconds (estimated)

nyh commented on test/cql-pytest/util.py:
```diff
@@ -139,6 +143,53 @@ def new_secondary_index(cql, table, column, name='', extra=''):
+    endpoint = cql.hosts[0].endpoint
+    session = cql_session(host=endpoint.address, port=endpoint.port, ssl=False, username=username, password=username)
+    yield session
+    session.shutdown()
```
Is session.shutdown() correct? I thought it needs to be cluster.shutdown() (remember that session is what is returned by cluster.connect())... Or maybe session.shutdown() does cluster.shutdown(), or maybe session.shutdown() is enough?

psarna replied:
session.shutdown() shuts down the session (https://docs.datastax.com/en/drivers/python/3.2/api/cassandra/cluster.html#cassandra.cluster.Session.shutdown), while cluster.shutdown() shuts down all the sessions. I guess in this case I should make the whole helper, cql_session, a contextmanager and kill the cluster in there. I'll rework it and send v2. Technically it probably still "just works", but it may leave some stray useless control connections laying around in memory, so it's better to be thorough.

nyh commented on test/cql-pytest/test_permissions.py:
```diff
@@ -39,3 +40,66 @@ def test_grant_applicable_data_and_role_permissions(cql, test_keyspace, cassandr
+        return
+
+# Test that data permissions can be granted and revoked, and that they're effective
+def test_udf_grant_revoke_data_permissions(cql, test_keyspace):
```
The name of this function seems to be wrong - since it's not related to udf or "data permissions". You can have one big test (or multiple smaller tests) for all the various types of permissions, and have the UDF test seperately (if I understand correctly, the UDF part will still fail because this feature isn't completely implemented?)

psarna replied:
oh, indeed, it has no connection to udf whatsoever; as for "data", it is related - in Scylla and Cassandra nomenclature, ALL KEYSPACES, KEYSPACE ks and TABLE t are all "data resources", as opposed to "role resources" and "function resources".

nyh replied:
Weird. In my mind "data" is the rows, the table itself is "metadata". But ok :-)

psarna commented:
v4:
 - migrated `cql_session` and `cql` helpers to contextmanagers, so that clusters and connections are properly shut down

psarna commented on test/cql-pytest/test_permissions.py:
```diff
@@ -39,3 +40,66 @@ def test_grant_applicable_data_and_role_permissions(cql, test_keyspace, cassandr
+            cql.execute(f"REVOKE CREATE ON ALL KEYSPACES FROM {username}")
+            def create_keyspace_idempotent():
+                user_session.execute(f"CREATE KEYSPACE {ks} WITH REPLICATION = {{ 'class' : 'SimpleStrategy', 'replication_factor' : 1 }}")
+                user_session.execute(f"DROP KEYSPACE {ks}")
```
oh I think the creator of any keyspace has implicit rights to also dropping it, in order to allow rolling back on something you did accidentally. I don't have any specification to quote, but that behavior is consistent in Scylla and Cassandra

psarna replied:
I just double-checked, and it works like this - once you create a keyspace or a table, as an author you "inherit" full rights over it, and these rights are simply present in the `system_auth.role_permissions` table.

psarna commented on test/cql-pytest/test_permissions.py:
```diff
@@ -39,3 +40,66 @@ def test_grant_applicable_data_and_role_permissions(cql, test_keyspace, cassandr
+                cql.execute(f"GRANT CREATE ON KEYSPACE {keyspace} TO {username}")
+                eventually_authorized(lambda: user_session.execute(f"CREATE TABLE {keyspace}.{t}(id int primary key)"))
+                # Permissions for a specific table
+                cql.execute(f"GRANT ALTER ON {keyspace}.{t} TO {username}")
```
right, I'll try to create a more useful idiom

psarna commented:
v5:
 - the test case is expanded to also check for inherited rights, e.g. that CREATE ON ALL KEYSPACES also grants one a right to create tables in any keyspace
 - some more checks for DROP permissions are added
 - one more helper function appeared, allowing to check if a certain function fails to run without permissions, and succeeds with permissions granted; it's still based on a lambda instead of a string statement, because sometimes it might be a series of statements, which is easier to express as a lambda

nyh commented:
> v5:
> 
>     * the test case is expanded to also check for inherited rights, e.g. that CREATE ON ALL KEYSPACES also grants one a right to create tables in any keyspace

I don't think this example is an "inherited right". Simply the Cassandra documentation (https://docs.datastax.com/en/cql-oss/3.x/cql/cql_reference/cqlGrant.html) says that "CREATE" permission on "ALL KEYSPACES" gives you the right to: "CREATE KEYSPACE and CREATE TABLE in any keyspace."

The DROP of a keyspace that you created is perhaps a less obvious case (that I still am not sure exactly how it works...).

nyh commented on test/cql-pytest/conftest.py:
```diff
@@ -36,34 +34,15 @@ def pytest_addoption(parser):
+        password="cassandra"
+    ) as session:
+        yield session
+        session.shutdown()
```
As I noted elsewhere (maybe you replied elsewhere), I don't understand if session.shutdown() is enough, or you need cluster.shutdown().
If you need cluster.shutdown(), you can make cql_session **itself** a contextmanager instead of a regular function, and the cql fixture can simply do
```
   yield from cql_session(....)
```
(yield from is a new Python feature, the older way to do it would be
```
with cql_session(...) as session:
   yield session
```

psarna replied:
Oh, but this shutdown() is just for completion. I also made the whole `cql_session` helper a contextmanager, and it properly calls `cluster.shutdown()` after the session is done (e.g. here) - you can see the change in util.py

nyh replied:
Oh, I missed that... Sorry, Ok. The session.shutdown() is not necessary because cluster.shutdown() does this already, but I guess it doesn't hurt....

nyh commented on test/cql-pytest/test_permissions.py:
```diff
@@ -39,3 +40,89 @@ def test_grant_applicable_data_and_role_permissions(cql, test_keyspace, cassandr
+# with given `permission` granted, and returns an unauthorized error
+# with the `permission` revoked.
+def check_enforced(cql, username, permission, resource, function):
+    grant(cql, permission, resource, username)
```
I suggested you could also check that it doesn't work before the grant. If I understand correctly, the user should not have any rights before the grant, right?

nyh reviewed (commented):
I think it's almost ready, but I'm still uneasy about the session.shutdown(), I think it should be cluster.shutdown() (I remember reading it in the past but I don't remember why). Because cql_session doesn't return cluster, I think it can be perhaps fixed by making cql_session a contextmanager instead of a function. But maybe there's another way (maybe the session points to the cluster it belongs to?)
UPDATE: I found what cluster.shutdown() does: http://datastax.github.io/python-driver/_modules/cassandra/cluster.html#Cluster.shutdown  - it calls session.shutdown() but also shuts down other control connections and other stuff. In the past, when I didn't shut those down, I got various warnings later when those things lingered in the background.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1275/

nyh commented:
It seems one of the tests failed in just one of the runs (https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1275/testReport/junit/(root)/test_permissions/test_grant_revoke_data_permissions/). I don't know why - a timeout?

psarna commented:
v6:
 - added another check to check_enforced - that a resource is not accessible also *before* the user is granted permissions to use it

nyh commented:
A test failed again. This time the problem is:
```python
                    eventually_unauthorized(lambda: user_session.execute(f"CREATE TABLE {keyspace}.{t}(id int primary key)"))
                    grant(cql, 'CREATE', f'KEYSPACE {keyspace}', username)
>                   eventually_authorized(lambda: user_session.execute(f"CREATE TABLE {keyspace}.{t}(id int primary key)"))
```
I think the problem here is that on the first line eventually_unauthorized eventually did NOT succeed, but it's possible that the first call actually did succeed, and then the table existed when the second line eventually_authorized() ran and failed.

nyh commented:
> Yes, you're right, this one isn't in fact idempotent. I'll rework it, and also remove one redundant permission check from the beginning, since it's now already checked by `check_enforced`.

I'm a bit worried that some of these tests appear to be (or at least, were in previous versions) "flaky", i.e., could sometime fail although there's no bug. If we add flaky tests to our test suite, this will annoy us in the future :-( Maybe you can run these tests manually N times (for some N >> 1) to perhaps find additional places where they're flaky?

psarna commented:
> > Yes, you're right, this one isn't in fact idempotent. I'll rework it, and also remove one redundant permission check from the beginning, since it's now already checked by `check_enforced`.
> 
> I'm a bit worried that some of these tests appear to be (or at least, were in previous versions) "flaky", i.e., could sometime fail although there's no bug. If we add flaky tests to our test suite, this will annoy us in the future :-( Maybe you can run these tests manually N times (for some N >> 1) to perhaps find additional places where they're flaky?

I do - I leave them in a loop for a while, both on Scylla and Cassandra. It's still just heuristics, but I can at least state that the tests passed on my machine multiple times in a row

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1304/

psarna commented:
Sigh, another silly mistake, I reused a function for "idempotent" table creation, but forgot that I also generated another unique name for it. Will fix in a few minutes

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1309/
